Fix banner rotation

charles-allen opened the pull request:
## ✅ DoD

- [x] All work is complete
- [x] Migrations: NA
- [x] Documentation: NA
- [x] Release notes: NA
- [x] Issues linked: resolves #41 

## 📝 Summary

Functional updates / fixes:
- Don't call resetTimer from the timed function!
- Don't call resetTimer if banners <= 1
- Hide prev/next buttons if no prev/next (banners <= 1)
- Blank banner during loading (to avoid layout jerk)
- Prev button clicks now wrap-around
- Exit `shiftSlide` early if <= 1 banner (shouldn't be possible)
- More realistic example data for `dateExpire`

Code style updates:
- Extract carousel to `HomeBanners.vue`
- Factor out CSS code updates to `BannerElement`
- `nextSlide`/`previousSlide` now share code  
    (by delegating to `shiftSlide`/`jumpToSlide`)

charles-allen committed:
[Banners] Fix rotation

- Don't call resetTimer from timer!
- Don't call resetTimer before we have data
- Don't start timer if banners <= 1
- Don't change classes if active === next
- Add wrap-around for prev button
- Hide prev/next if none (banners <= 1)
- More realistic example `dateExpire`
- No `let`
- Comments

charles-allen committed:
[Banners] Short-circuit when banners <2

charles-allen committed:
Revert "[Test-Data] Move data to sub-folder"

This reverts commit 5622336ec323aabfdef4628e22ad33b87edbd156.

charles-allen committed:
[Test-Data] Move data to sub-folder

...streamline script

charles-allen committed:
[Banners] Fix multi-method @click

Nutto55 commented on hosting/src/global/Home/Home.html:
```diff
@@ -15,14 +15,16 @@
+      v-if="banners.length > 1"
       class="absolute bottom-0 left-0 rounded-tr-md sm:rounded-tr-xl lg:rounded-tr-3xl bg-black bg-opacity-30 pr-1 w-8 h-8 sm:w-10 sm:h-10 md:w-12 md:h-12 lg:w-14 lg:h-14 text-lg sm:text-xl md:text-2xl lg:text-4xl flex items-center justify-center text-white cursor-pointer"
-      @click="previousSlide"
+      @click="previousSlide(); resetSlideTimer()"
```
This look cool! I never know that we can trigger on 2 functions like this! wow

charles-allen replied:
Neither did I. If you look at the earlier commit, I tried to do it with `&&` (I must have been thinking shell), and I didn't immediately notice that it wasn't working!

Nutto55 commented on hosting/src/global/Home/Home.ts:
```diff
@@ -29,62 +29,98 @@ export default class Home extends Vue {
+    if (activeSlide === null) return // no slides
+
+    const previousSlide = activeSlide.previousElementSibling
+    if (previousSlide === null || !previousSlide.classList.contains('slide')) {
+      // Return to end...
+      const allSlides = activeSlide.parentElement?.querySelectorAll('.slide') ?? []
+      const lastSlide = allSlides.length > 0 ? allSlides[allSlides.length - 1] : null
+      if (lastSlide === null) return
+
+      // Old => Exit to left
+      activeSlide.classList.remove('translate-x-0')
+      activeSlide.classList.add('-translate-x-full')
+
+      // New => Enter from right
+      lastSlide.classList.remove('translate-x-full')
+      lastSlide.classList.add('translate-x-0')
+
+      // Waiting => Ready to move on from left
+      const otherSlides = document.querySelectorAll('.slide.translate-x-full')
+      otherSlides.forEach(element => {
+        element.classList.remove('translate-x-full')
+        element.classList.add('-translate-x-full')
+      })
+    } else {
+      // Old => Exit to right
+      activeSlide.classList.remove('translate-x-0')
+      activeSlide.classList.add('translate-x-full')
+
+      // New => Enter from left
+      previousSlide.classList.remove('-translate-x-full')
+      previousSlide.classList.add('translate-x-0')
+    }
```
I think you can add `return` after if case (after for each loop). So, you don't need the last `else`. you can just attach

```
// Old => Exit to right
activeSlide.classList.remove('translate-x-0')
activeSlide.classList.add('translate-x-full')

// New => Enter from left
previousSlide.classList.remove('-translate-x-full')
previousSlide.classList.add('translate-x-0')
```
After that if case with `return`.

charles-allen replied:
I hadn't considered that. I inherited this structure from @antonyharfield. 

I think I prefer to limit early returns to guards. This if block is significantly longer. The else communicates clearly that the bottom 4 lines only execute on 1 branch.

That said... if we invert the if, we can simplify the condition, and move the shorter path first:
```ts
if (previousSlide?.classList?.contains('slide')) ...
```

I might also consider creating methods like:
```ts
slideInRight(slide: Element): boolean {
  slide.classList.remove('-translate-x-full')
  slide.classList.add('translate-x-0')
  return true
}

slideOutRight(slide: Element): boolean {
  slide.classList.remove('translate-x-0')
  slide.classList.add('translate-x-full')
  return true
}
```

Then perhaps it could be more guard-like:
```ts
if (...) return slideOutLeft(activeSlide) && slideInLeft(previousSlide)
```

charles-allen replied:
Probably there is also some repetition between `nextSlide` & `previousSlide`; the "wrap-around" logic of one method is similar to the standard logic of the other function. Extracting `slideXXX` to methods removes that repetition & decreases the chance of making edits in on one function (just earlier today I accidentally patched one and not the other).

Nutto55 replied:
> if (...) return slideOutLeft(activeSlide) && slideInLeft(previousSlide)

This also look cool!

I don't have many experience on the banner(I think it is/will look like carousal?). Normally, I do the logic by checking the index of the current banner, so it will slide forward and backward. I do that by create 2 function e.g. back() and next(), then if it is last (checking by array length) then just rolling back to first banner again like looping.

charles-allen replied:
It turns out the if is *really* ugly with our lint rules:
```ts
if(previousSlide && previousSlide.classList.contains('slide')) // lint prefers optional-chaining
if(previousSlide?.classList?.contains('slide') ?? false) // TS couldn't infer that slide isn't null
if(previousSlide?.classList?.contains('slide') === true) // ewww
```

Yes, it's a carousel.

Yes, I think holding an array of "slide" elements & the current index is easier. I'm going to refactor to that, using `ref`. I might also wrap the HTML elements in a class with methods like `slideToLeft()`, `slideToMiddle()`, etc.

And I think it's finally time to extract the carousel to its own component.

charles-allen replied:
I think I should also add a new section to our PR template... "what is this feature? (background/introduction)" to make it easier for reviewers to understand the PR.

charles-allen replied:
Replaced this code; see [new discussion](https://github.com/phits-tech/website/pull/43/files#r628838693)

charles-allen committed:
[Banners] Extract to separate component

charles-allen committed:
[Banners] Capture Elements using `ref`

charles-allen committed:
[Banners] BannerElement wrapper

- BannerElement encapsulates `moveToXXX()`
- Merged next/previous code as `shiftSlide(...)`
- Can also jump to arbitrary slides (untested)

charles-allen committed:
[Banners] Add loading banner

...to stop layout jerk

charles-allen committed:
[Spaces] Delete obsolete images

charles-allen commented on hosting/src/global/Home/HomeBanners.ts:
```diff
@@ -0,0 +1,91 @@
+    if (this.bannerSlides.length > 1) this.nextSlideInterval = setInterval(this.nextSlide, 5000)
+  }
+
+  nextSlide(): void { this.shiftSlide(1) }
+  previousSlide(): void { this.shiftSlide(-1) }
+
+  shiftSlide(shift: -1 | 1): void {
+    if (this.bannerSlides.length < 2) return
+
+    // Detect wrap-around
+    const newIndex = this.currentSlideIndex + shift
+    const newIndexCorrected = (newIndex < 0) ? this.bannerSlides.length - 1 : (newIndex > this.bannerSlides.length - 1) ? 0 : newIndex
+    if (newIndex !== newIndexCorrected) return this.jumpToSlide(newIndexCorrected)
+
+    // Shift once
+    if (shift === 1) this.bannerSlides[this.currentSlideIndex].moveToLeft()
+    else this.bannerSlides[this.currentSlideIndex].moveToRight()
+    this.bannerSlides[newIndexCorrected].moveToCenter()
+    this.currentSlideIndex = newIndexCorrected
+  }
+
+  jumpToSlide(newIndexCorrected: number): void {
+    this.bannerSlides.forEach((slide, idx) => {
+      if (idx > newIndexCorrected) slide.moveToRight()
+      else if (idx < newIndexCorrected) slide.moveToLeft()
+      else slide.moveToCenter()
+    })
+    this.currentSlideIndex = newIndexCorrected
+  }
```
@Nutto55 - I've done a major refactor based on [our discussion](https://github.com/phits-tech/website/pull/43#discussion_r628737080).

In the new code:
- I extracted CSS updates to a class `BannerElement`  
    (methods: `moveToLeft`, `moveToCenter`, `moveToRight`)
- I merged next & prev into a new function `shiftSlide`
- `shiftSlide` detects if a shift would wrap-around, and passes those case to `jumpToSlide`

...actually in writing this, I think I've seen an optimization! BRB ;p

charles-allen replied:
Done :) I realized I could eliminate the ugly ifs here, by adding a `move(shift)` method to `BannerElement`.

charles-allen committed:
[Banners] Encapsulate shift direction

charles-allen committed:
[Banners] Remove obsolete styles

charles-allen committed:
[Cleaning] Add some comments

charles-allen commented on hosting/src/global/Home/HomeBanners.ts:
```diff
@@ -0,0 +1,93 @@
+  jumpToSlide(newIndexCorrected: number): void {
+    this.bannerSlides.forEach((slide, idx) => slide.move(newIndexCorrected - idx))
+    this.currentSlideIndex = newIndexCorrected
+  }
```
@Nutto55 - I've done a major refactor based on [our discussion](https://github.com/phits-tech/website/pull/43#discussion_r628737080).

In the new code:
- I extracted CSS updates to a class `BannerElement` (`move`, `moveToLeft`, `moveToCenter`, `moveToRight`)
- I merged to logic of `nextSlide` & `previousSlide` as a new function `shiftSlide`
- `shiftSlide` detects if a shift would wrap-around, and passes those case to `jumpToSlide`

charles-allen commented on hosting/src/global/Home/HomeBanners.ts:
```diff
@@ -0,0 +1,100 @@
+    if (this.bannerSlides.length > 1) this.nextSlideInterval = setInterval(this.nextSlide, 5000)
+  }
+
+  nextSlide(): void { this.shiftSlide(1) }
+  previousSlide(): void { this.shiftSlide(-1) }
+
+  shiftSlide(shift: -1 | 1): void {
+    if (this.bannerSlides.length < 2) return
+
+    // Detect wrap-around
+    const newIndex = this.currentSlideIndex + shift
+    const newIndexCorrected = (newIndex < 0) ? this.bannerSlides.length - 1 : (newIndex > this.bannerSlides.length - 1) ? 0 : newIndex
+    if (newIndex !== newIndexCorrected) return this.jumpToSlide(newIndexCorrected)
+
+    // Shift once
+    this.bannerSlides[this.currentSlideIndex].move(shift)
+    this.bannerSlides[newIndexCorrected].moveToCenter()
+    this.currentSlideIndex = newIndexCorrected
+  }
+
+  jumpToSlide(newIndexCorrected: number): void {
+    // slideIdx < current => wait on left
+    // slideIdx == current => move to center
+    // slideIdx > current => wait on right
+    this.bannerSlides.forEach((slide, slideIdx) => slide.move(newIndexCorrected - slideIdx))
+    this.currentSlideIndex = newIndexCorrected
+  }
```
@Nutto55 - I've done a major refactor based on [our discussion](https://github.com/phits-tech/website/pull/43#discussion_r628737080). In the new code:
- I extracted CSS updates to a class `BannerElement` (`move`, `moveToLeft`, `moveToCenter`, `moveToRight`)
- I merged to logic of `nextSlide` & `previousSlide` as a new function `shiftSlide`
- `shiftSlide` detects if a shift would wrap-around, and passes those case to `jumpToSlide`  
  (jumping to a slide / wrap-around requires updating all the slides)

antonyharfield replied:
I'm glad you have made sense of this code... apologies the first version was so buggy!

charles-allen replied:
I wasn't really planning to refactor it; just fix the edge-cases. But @Nutto55 kept bugging me that it could be cleaned up :p

charles-allen committed:
[Banners] Longer pause after user interaction

charles-allen committed:
[Banners] Move data to Vuex

charles-allen committed:
[Banners] Encapsulate constants

charles-allen committed:
Merge branch 'develop' into feature/banner-rotation

charles-allen committed:
Merge branch 'develop' into feature/banner-rotation
